content/specs/event-stream.md

```diff
@@ -0,0 +1,133 @@
+
+- `405 Method Not Allowed`: Returned to client for non-GET HTTP requests to a stream endpoint.
+- `426 Upgrade Required`: Returned to client if if `Upgrade` header is not included in a request to a stream endpoint.
+- `429 Too Many Requests`: Frequently used for rate-limiting. Client may try again after a delay. Support for the `Retry-After` header is encouraged.
+- `500 Internal Server Error`: Client may try again after a delay
+- `501 Not Implemented`: Service does not implement WebSockets or streams, at least for this endpoint. Client should not try again.
+- `502 Bad Gateway`, `503 Service Unavailable`, `504 Gateway Timeout`: Client may try again after a delay
+
+Servers *should* return HTTP bodies as JSON with the standard XRPC error message schema for these status codes. But clients also need to be robust to unexpected response body formats. A common situation is receiving a default load-balancer or reverse-proxy error page during scheduled or unplanned downtime.
+
+Either the server or the client may decided to drop an open stream connection if there have been no messages for some time. It is also acceptable to leave connections open indefinitely.
+
+### Framing
+
+Each binary WebSocket frame contains two DAG-CBOR objects, concatenated. The first is a "header" and the second is the "payload".
+
+The header DAG-CBOR object has the following fields:
+
+- `op` ("operation", integer, required): fixed values, indicating what this frame contains
+    - `1`: a regular message, with type indicated by `t`
+    - `-1`: an error message
+- `t` ("type", string, optional): required if `op` is `1`, indicating the Lexicon sub-type for this message, in short form. Does not include the full Lexicon identifier, just a fragment. Eg: `#commit`. Should not be included in header if `op` is `1`
+
+Clients should ignore frames with headers that have unknown `op` or `t` values. Unknown fields in both headers and payloads should be ignored. Invalid framing or invalid DAG-CBOR encoding are hard errors, and the client should drop the entire connection instead of skipping the frame. Servers should ignore any frames received from the client, not treat them as errors.
+
+Error payloads all have the following fields:
+
+- `error` (string, required): the error type name, with no namespace or `#` prefix
+- `message` (string, optional): a description of the error
+
+Streams should be closed immediately following transmitting or receiving an error frame.
+
+Message payloads must always be objects. They should omit the `$type` field, as this information is already indicated in the header. There is no specific limit on the size of WebSocket frames in atproto, but they should be kept reasonably small. A couple megabytes is an informal guideline.
+
+If a client can not keep up with the rate of messages, the server may send a "too slow" error and close the connection.
+
+### Sequence Numbers
```

Comment:
Oof, this is a lot. Sure would be nice if we could get some of it for free by switching the namespace from PDS hostname + endpoint to user DID + collection and then depending on monotonically increasing TIDs instead. I guess that would mean servers and clients would have to keep track of O(users) sequence TIDs instead of just O(servers/clients + endpoints) sequence numbers though.

Reply:
This is under pretty active discussion internally, and with a few external tech advisors. Hard to compress all the conversation to a short summary. Also, realistically, this is fairly high up in the protocol stack and the sort of thing that might have multiple implementations over time. Eg, internally and at scale, can't really imagine not using Kafka for the big firehoses; maybe we'll have a formal "how to do this in Kafka within an org/consortium" spec or recommendation.

Reply:
Also, clarified that the sequence stuff is optional. Can totally create subscription Lexicons which don't include it.